File: array-string/text-justification/text-justification.ts
function fullJustify(words: string[], maxWidth: number): string[] {
    const res: string[] = [];
    let i = 0;

    while (i < words.length) {
        let j = i, lineLen = 0;
        while (j < words.length && lineLen + words[j].length + (j - i) <= maxWidth) {
        lineLen += words[j].length;
        j++;
        }

        const gaps = j - i - 1;
        const lastLine = j === words.length;

        if (lastLine || gaps === 0) {
        const line = words.slice(i, j).join(" ");
        res.push(line + " ".repeat(maxWidth - line.length));
        } else {
        const totalSpaces = maxWidth - lineLen;
        const spaceEach = Math.floor(totalSpaces / gaps);
        const extra = totalSpaces % gaps;

        let line = "";
        for (let k = i; k < j; k++) {
            line += words[k];
            if (k < j - 1) {
            const spaces = spaceEach + (k - i < extra ? 1 : 0);
            line += " ".repeat(spaces);
            }
        }
        res.push(line);
        }

        i = j;
    }

    return res;
};

File: array-string/text-justification/text-justification.go
func fullJustify(words []string, maxWidth int) []string {
	res := []string{}
	for i := 0; i < len(words); {
		j, lineLen := i, 0
		for j < len(words) && lineLen+len(words[j])+(j-i) <= maxWidth {
			lineLen += len(words[j])
			j++
		}

		gaps := j - i - 1
		lastLine := j == len(words)

		if lastLine || gaps == 0 {
			line := strings.Join(words[i:j], " ")
			if len(line) < maxWidth {
				line += strings.Repeat(" ", maxWidth-len(line))
			}
			res = append(res, line)
		} else {
			totalSpaces := maxWidth - lineLen
			spaceEach := totalSpaces / gaps
			extra := totalSpaces % gaps

			var b strings.Builder
			b.Grow(maxWidth)
			for k := i; k < j; k++ {
				b.WriteString(words[k])
				if k < j-1 {
					spaces := spaceEach
					if k-i < extra {
						spaces++
					}
					b.WriteString(strings.Repeat(" ", spaces))
				}
			}
			res = append(res, b.String())
		}

		i = j
	}
	return res
}